metals/src/main/scala/scala/meta/internal/metals/TastyHandler.scala

```diff
@@ -0,0 +1,151 @@
+    languageClient: MetalsLanguageClient,
+    clientConfig: ClientConfiguration,
+    httpServer: () => Option[MetalsHttpServer]
+)(implicit ec: ExecutionContext) {
+
+  def executeShowTastyCommand(
+      params: l.ExecuteCommandParams
+  ): Future[Unit] =
+    if (
+      clientConfig.isExecuteClientCommandProvider() && !clientConfig
+        .isHttpEnabled()
+    ) {
+      val tastyResponse = getTasty(params)
+      tastyResponse.map { tasty =>
+        val command = new l.ExecuteCommandParams(
+          "metals-show-tasty",
+          List[Object](tasty).asJava
+        )
+        languageClient.metalsExecuteClientCommand(command)
+        scribe.debug(s"Executing show TASTy ${command}")
+      }
+    } else {
+      (httpServer(), parseJsonParams(params)) match {
+        case (Some(server), Some(uri)) =>
+          Future.successful(
+            Urls.openBrowser(server.address + s"/tasty?file=$uri")
+          )
+        case (None, _) =>
+          Future.successful {
+            scribe.warn(
+              "Unable to run show tasty. Make sure `isHttpEnabled` is set to `true`."
+            )
+          }
+        case other =>
+          scribe.error(s"Show TASTy command failed $other")
+          languageClient.showMessage(
+            Messages.showTastyFailed
+          )
+          Future.successful(())
+      }
+    }
+
+  def getTastyForURI(uri: URI): Future[String] =
+    getTasty(
+      uri,
+      clientConfig.isCommandInHtmlSupported(),
+      clientConfig.isHttpEnabled()
+    )
+
+  private def getTasty(
+      params: l.ExecuteCommandParams
+  ): Future[String] =
+    parseJsonParams(params) match {
+      case Some(path) =>
+        val uri = new URI(path)
+        getTastyForURI(uri)
+      case None =>
+        val error = s"Error, invalid show TASTy arguments $params"
+        scribe.error(error)
+        Future.successful(error)
+    }
+
+  private def parseJsonParams(
+      commandParams: l.ExecuteCommandParams
+  ): Option[String] = for {
+    args <- Option(commandParams.getArguments)
+    arg <- args.asScala.headOption.collect {
+      case js: JsonPrimitive if js.isString => js
+    }
+  } yield arg.getAsString()
+
+  /**
+   * For given uri (could be both .scala and .tasty file) try to find:
+   * - build target
+   * - uri to existing tasty file
+   * Iff they both exists get decoded tasty file content (pc.getTasty takes into account if client supports html)
+   */
+  private def getTasty(
+      uri: URI,
+      isHtmlSupported: Boolean,
+      isHttpEnabled: Boolean
+  ): Future[String] = {
+    val absolutePathOpt = Try(AbsolutePath.fromAbsoluteUri(uri)) match {
+      case Success(value) => Some(value)
+      case Failure(exception) =>
+        scribe.error(exception.toString())
+        None
+    }
+    val pcAndTargetUriOpt = for {
+      absolutePath <- absolutePathOpt
+      buildTargetId <- buildTargets.inverseSources(absolutePath)
+      buildTarget <- buildTargets.scalaTarget(buildTargetId)
```

Comment:
I saw you asked a question about letting the user know that it's not supported in Scala 3. I think it'd be nice to just warn them in that situation with a message. We can check that here once we have the `buildTarget`.

Reply:
I added here the check, but IMO differentiating between warning and error might add unnecessary complexity to the code since it would require a third possible result. Let's call it the da. 
If users would start to complain about this, then we could alter it. Seems fine?

Reply:
👍🏼 sounds good to me.